### DISCUSSION TITLE
Creating custom Controls (Is there a better way?)

### DISCUSSION THREAD

**Discussion by MDale-RAC** [CommunityToolkit/Maui#1183] [Q&A] [answered]
When creating a `ContentView` is there a more concise way to add `BindableProperty`s?

For instance (in the Code-Behind):
```C#
public partial class ProfileDetailView: ContentView
{
	public static readonly BindableProperty EmailProperty =
		BindableProperty.Create(nameof(Email), typeof(string), typeof(LoginView), string.Empty);

	public string Email
	{
		get => (string)GetValue(EmailProperty);
		set => SetValue(EmailProperty, value);
	}

	[...]
}
```

Seems, to me, incredibly verbose when compared to what is available in `ObservableObject` (for ViewModels) using the `ObservablePropertyAttribute`:

```C#
[ObservableProperty]
private string _email = string.Empty;
```

I understand these are different paradigms for different use-cases, but is there a way to generate the above (`BindableProperty` and `get`ter/`set`ter) with a similar type of attribute? For instance, say (naively):

```C#
public partial class ProfileDetailView: ContentView
{
	[BindableProperty]
	private string _email = string.Empty;
	// Auto-generate:
	// 	|- public static readonly BindableProperty EmailProperty
	//		\- propertyName: Email
	//		  \- returnType: typeof(string)
	//		    \- declaringType: typeof(ProfileDetailView)
	//		      \- defaultValue: string.Empty (?? this I am dubious/unsure of ??)
	// 	|- public string Email
	//		\- getter (cast GetValue of EmailProperty)
	//		  \- setter (SetValue for EmailProperty using value)

	[...]
}
```

Does this already exist perhaps? (If so, would someone be so kind as to direct me to that documentation? I didn't see it, but perhaps I don't know the right term for it!)

Many thanks in advance!

**Comment by pictos** [comment]
There're nuget packages that do it for you, we have an open spec for that but isn't started yet.

**Reply by MDale-RAC** [reply]
Sorry, but your reply is ambiguous to me:
- "[There are] nuget packages" (what nuget packages?)
- "open spec for that" (what is "that"?)

What keywords would you suggest looking for?

Any chance you could provide any links to what you are referencing?

**Reply by IeuanWalker** [reply, marked as answer]
@MDale-RAC the spec is the potential design for it, for people to discuess how it should be implemented before someone starts to work on it - https://github.com/CommunityToolkit/Maui/issues/542

There are existing nuget packages that you could use, such as - https://github.com/rrmanzano/maui-bindableproperty-generator

**Reply by bijington** [reply]
@IeuanWalker thanks you beat me to it!

**Comment by CliffAgius** [comment]
Closed as answered
